x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceStats.java

```diff
@@ -263,6 +289,15 @@ public Accumulator incFailure() {
             }
         }
 
+        public Accumulator incCacheMiss() {
+            readWriteLock.readLock().lock();
```

Comment:
Not something for this PR, but I wonder if the read/write lock is over complication.  The things that are going on inside the locked sections of code are very simple operations.  In my experience you need to be doing a _lot_ of stuff in the locked sections to make the cost of a read/write lock worthwhile.  The amount of work the lock is doing to ensure fairness probably dwarfs the amount of work in these methods.  Then each `LongAdder` is doing its own synchronization internally too.  It is probably the case that it would be more efficient to simply make all the methods that use the read/write lock `synchronized`, replace the `LongAdder`s with simple `long`s, and just use `++` to increment them.

Reply:
I will do some measurements and see which is better :)

Reply:
Yet again, it is better to measure before making a decision about "what is faster". 

Here are average microsecond timed benchmarks for 1 and 128 threads. 

example code:
```
for (int i = 1; i < 100_001; ++i) {
    bh.consume(accumulator.incMissingFields().incFailure().incInference());
}
```

The scenario was the BEST cast scenario for the RW lock (only ever doing increments, and never resetting with a full write lock).
```Benchmark                                                   (interrupts)  Mode  Cnt        Score        Error  Units
MultiThreadedStatsAccumulatorBenchmark.rwAccumulator_1             10000  avgt   20     5957.399 ±    112.892  us/op
MultiThreadedStatsAccumulatorBenchmark.rwAccumulator_128           10000  avgt   20  7480921.908 ± 255364.820  us/op
MultiThreadedStatsAccumulatorBenchmark.syncAccumulator_1           10000  avgt   20      421.662 ±      2.616  us/op
MultiThreadedStatsAccumulatorBenchmark.syncAccumulator_128         10000  avgt   20   792910.927 ±  52219.577  us/op
```

full benchmark code: https://gist.github.com/benwtrent/b74254bf64aba0c55d4b24a474f9eff1

@droberts195 ^

Reply:
Yes, that tallies with what I've always seen in the past, which is that you have to do a lot of work in the locked sections for the benefit of an RW lock to outweigh the complexity of its internal code.

Up to you if you want to change it in this PR or a different one though.